inter_company_rules/models/account_invoice.py

```diff
@@ -0,0 +1,230 @@
+        journal = self.env['account.journal'].search([
+            ('type', '=', journal_type),
+            ('company_id', '=', company.id)
+        ], limit=1)
+        if not journal:
+            raise UserError(_(
+                'Please define %s journal for this company: "%s" (id:%d).')
+                % (journal_type, company.name, company.id))
+
+        # find periods of supplier company
+        context = self._context.copy()
+        context['company_id'] = company.id
+        period_ids = self.env['account.period'].with_context(context).find(
+            self.date_invoice)
+
+        # find account, payment term, fiscal position, bank.
+        partner_data = self.onchange_partner_id(inv_type,
+                                                self.company_id.partner_id.id,
+                                                company_id=company.id)
+        if not self.currency_id.company_id:
+            # currency shared between companies
+            currency_id = self.currency_id.id
+        else:
+            # currency not shared between companies
+            curs = self.env['res.currency'].with_context(context).search([
+                ('name', '=', self.currency_id.name),
+                ('company_id', '=', company.id),
+                ], limit=1)
+            if not curs:
+                raise UserError(_(
+                    "Could not find the currency '%s' in the company '%s'")
+                    % (self.currency_id.name, company.name_get()[0][1]))
+            currency_id = curs[0].id
+        return {
+            'name': self.name,
+            # TODO : not sure !!
+            'origin': self.company_id.name + _(' Invoice: ') + str(
+                self.number),
+            'supplier_invoice_number': self.number,
+            'check_total': self.amount_total,
+            'type': inv_type,
+            'date_invoice': self.date_invoice,
+            'reference': self.reference,
+            'account_id': partner_data['value'].get('account_id', False),
+            'partner_id': self.company_id.partner_id.id,
+            'journal_id': journal.id,
+            'currency_id': currency_id,
+            'fiscal_position': partner_data['value'].get(
+                'fiscal_position', False),
+            'payment_term': partner_data['value'].get('payment_term', False),
+            'company_id': company.id,
+            'period_id': period_ids and period_ids[0].id or False,
+            'partner_bank_id': partner_data['value'].get(
+                'partner_bank_id', False),
+            'auto_generated': True,
+            'auto_invoice_id': self.id,
+        }
+
+    @api.model
+    def _prepare_invoice_line_data(self, line_data, line):
+        """ Generate invoice line values
+            :param line_data : dict of invoice line data
+            :rtype line_data : dict
+            :param line : the invoice line object
+            :rtype line : account.invoice.line record
+        """
+        vals = {
```

Comment:
Better to call the corresponding onchange method. The only concern here is if it will be correctly called in multi-company context.